py3status/modules/playerctl.py

```diff
@@ -0,0 +1,272 @@
+
+    def _init_player(self, player_name):
+        if not self._player_should_be_tracked(player_name):
+            return
+
+        player = Playerctl.Player.new_from_name(player_name)
+
+        # Initialize player event listeners
+        player.connect("playback-status", self._status_changed, self.manager)
+        player.connect("metadata", self._on_metadata, self.manager)
+
+        self.manager.manage_player(player)
+
+    def _start_loop(self):
+        loop = GLib.MainLoop()
+
+        # The loop is blocking so it needs to be run a separate thread
+        thread = Thread(target=loop.run)
+        thread.daemon = True
+        thread.start()
+
+    def _on_name_appeared(self, manager, player_name):
+        self._init_player(player_name)
+
+    def _on_name_vanished(self, manager, name):
+        # Add a very small delay to give Playerctl time to remove the player
+        time.sleep(0.01)
+        self.py3.update()
+
+    def _on_metadata(self, player, metadata, manager):
+        self.py3.update()
+
+    def _status_changed(self, player, status, manager):
+        self.py3.update()
+
+    @staticmethod
+    def _microseconds_to_time(microseconds):
+        seconds = microseconds // 1_000_000
+        m, s = divmod(seconds, 60)
+        h, m = divmod(m, 60)
+        time = f"{h}:{m:02d}:{s:02d}"
+        return time.lstrip("0").lstrip(":")
+
+    def _get_player_position(self, player):
+        try:
+            # Playerctl doesn't support getting positions for all players
+            microseconds = player.get_position()
+            position = self._microseconds_to_time(microseconds)
+        except GLib.GError:
+            position = None
+
+        return position
+
+    def _set_data_from_metadata(self, player, data):
+        """Set any data retrieved directly from the metadata for a player"""
+        metadata = dict(player.props.metadata)
+
+        duration_ms = metadata.get("mpris:length")
+        if duration_ms:
+            data["duration"] = self._microseconds_to_time(duration_ms)
+        else:
+            data["duration"] = None
+
+    def _get_player_data(self, player):
+        data = {}
+
+        # Song attributes
+        data["album"] = player.get_album()
+        data["artist"] = player.get_artist()
+        data["title"] = player.get_title()
+        data["position"] = self._get_player_position(player)
+        self._set_data_from_metadata(player, data)
+
+        # Player attributes
+        data["player"] = player.props.player_name
+        data["status"] = player.props.status.lower()
+
+        return data
+
+    def playerctl(self):
+        tracked_players = self.manager.props.players
+
+        # Only have a cache timeout if the user wants to know the players' positions
+        if tracked_players and (
+            self.py3.format_contains(self.format, "position")
+            or self.py3.format_contains(self.format_player, "position")
+        ):
+            cached_until = 1
```

Comment:
Sounds good. Is there a reason `cache_timeout` can't just be declared at the top of the class by the button and format default configurations instead of using `getattr`?

Reply:
Because once you expose a config, it can be hard to carry a config over to new one. This is more for `format_` related configs...

I think it would good to give users an impression that playerctl module's updates are decided automatically based on data changes/actions rather than at interval. If we expose this, then users may think this is done at interval.